### PR TITLE
Grammar rule refactoring and cross-linking

### PR DESCRIPTION
**Pull request by ingydotnet**
This PR makes the rule refs in grammar productions link to their rules in the
HTML.

It also stores the grammar rules outside spec.md in grammar/rules.yaml.
They are edited in the yaml file and the build process copies then back in
spec.md.

### CODE REVIEW

**Review by eemeli — COMMENTED** [review]
I think I'm still missing the point of this change -- why is it a good idea to have/keep the rules in an external doc? And why is that external file then the source of truth for the rules?

I would presume that it'd be easier overall if the source of truth was `spec.md`, with `rules.yaml` automatically generated from it. Why can't we do that?

As is, there seems to be a danger that we're relying on human oversight to make sure that no changes get applied to `spec.md` without a corresponding change to `rules.yaml`, as there are plenty of such changes (removing rules, changing rule contents) that would not show up in automation.

**Comment by ingydotnet** [comment]
TL;DR skip to last paragraph :)

@eemeli To start making spec changes that define the next version(s) of YAML we will need to restructure the grammar quite a bit. With every grammar change we generate parsers programatically and test them against the test suite (which should be the final SoT for defining YAML).

On top of that, as we start speccing features/rules, these will add or modify productions in the grammar. iow, a feature is a merge of alternate productions into the grammar.

The `spec.md` document will serve as explanation of the various parts of YAML, not the SSoT that it is now. The "spec" will be a combination of a core grammar, feature grammar overrides, a test suite, and many docs including `spec.md`.

Keeping the current `spec.md` as the SSoT of the grammar productions, is much too cumbersome to begin reorganizing the YAML grammar. IMHO trying to cram every detail defining the language into one long doc did not serve the language well. The grammar always should have been defined separately and a set of docs describing it and other concepts should have been written in a style that was easier to read. These docs would have have links to grammar rules and supporting tests as they made sense to read; not as a set of explanations around the entire grammar in order.

**All that said**, if you are more comfortable with simply extracting the grammar rules from `spec.md` for now so that grammar work can begin separate from doc work, I'm ok with that. I think that as we move forward it will become obvious to both of us where the various SSoTs belong...

**Comment by ingydotnet** [comment]
@eemeli I'll go ahead and adjust things to just extract the grammar rules from the `spec.md` SSoT for now...
